qnsolver/qnsolver.1
.\" -*- nroff -*-
.TH qnsolver 1 ""  "5.29"
.\" $Id: qnsolver.1 17172 2024-04-07 11:08:12Z greg $
.\"
.\" --------------------------------
.SH "NAME"
qnsolver \- Solve a conventional BCMP queueing network model using Mean Value Analysis
.SH "SYNOPSIS"
.br
.B qnsolver
[\fIOPTION \&.\|.\|.\fP] [\fIFILE\fR \&.\|.\|. ]
.SH "DESCRIPTION"
\fBqnsolver\fP is used to conventional BCMP queueing network models
using Exact or Approximate Mean Value Analysis.  The input file format
is that of the Java Modeling Tools (with extensions to handle the SPEX
modeling languge used by lqns(1).)  If a ``What-If'' element is found
in the input file, output is either CSV, or a self-contained file that
can be used with gnuplot(1) to plot the results.  Otherwise, the
results for the model are printed.
.SH "OPTIONS"
.TP
\fB\-b\fR, \fB\-\-bounds\fR
Use the Bounds solver to solve the model(s).  Bounds are computed
implicitly with \fB\-\-throughput\fR and
\fB\-\-response\-time\fR.
.TP
\fB\-b\fR, \fB\-\-no\-bounds\fR
Suppress bounds plotting for system throughput and system response
time output.
.TP
\fB\-B\fR, \fB\-\-export-jaba\fR
Export a model that can be used by the bounds solver from the Java
Modelling Tools.
.TP
\fB\-C\fR, \fB\-\-colours=ARG\fR
Use the comma separated list of colours specified in \fIarg\fP for
plotting. (not implemented at the moment).
.TP
\fB\-d\fR, \fB\-\-debug\-mva\fR
Output debug information from the MVA solver.  The MVA library needs
to be compiled with debugging enabled to output more than just summary information.
.TP
\fB\-e\fR, \fB\-\-exact\-mva\fR
Use Exact Mean Value Analysis to solve the model(s).
.TP
\fB\-f\fR, \fB\-\-fast\-linearizer\fR
Use ``Fast Linearizer'' approximate MVA to solve the model(s).
.TP
\fB\-J\fR, \fB\-\-export-jmva\fR
Export a model that can be used by the MVA solver from the Java
Modelling Tools.
.TP
\fB\-l\fR, \fB\-\-linearizer\fR
Use Linearizer approximate MVA to solve the model(s).
.TP
\fB\-L\fR, \fB\-\-debug\-lqx\fR
Debug the LQX program.
.TP
\fB\-m\fR, \fB\-\-multiserver\fR=\fIarg\fR
Use the multiserver solution specified by \fIarg\fP.  \fIArg\fP is one
of \fIconway\fP, \fIreiser\fP, \fIrolia\fP, or \fIzhou\fP.  The Zhou
and Rolia solutions should only be used with approximate MVA.  The
Reiser solution should only be used with exact MVA.
.TP
\fB\-M\fR, \fB\-\-force-multiserver\fR
Use the multiserver solution at all load-independent stations using a
mutliplicity of one.  Normally, the simpler load-independent soltions
are used instead.
.TP
\fB\-n\fR, \fB\-\-no\-execute\fR
Load the model and run any LQX (``What-If'') code.  Do not solve.
.TP
\fB\-o\fP, \fB\-\-output\fR=\fIarg\fR
Direct analysis results to \fIoutput\fP.  A filename of `\fI-\fR'
directs output to standard output.
.TP
\fB\-O\fR, \fB\-\-file\-format\fR=\fIarg\fR
For plotting, the default is to display the plot output to the
terminal for the host operating system.  This option add gnuplot code to
output the plotting instructions to a file using the format specified
in \fIarg\fP.  \fIArg\fP can be one of \fIemf\fP, \fIeps\fP, \fIfig\fP,
\fIgif\fP, \fIlatex\fP, \fIpdf\fP, \fIpng\fP, \fIsvg\fP.  The
output file name for the gnuplot code is the base name of the input
file being solved appended with the file format and with a suffix of \fI.gnuplot\fP.
.TP
\fB\-P\fR, \fB\-\-print\-lqx\fR
Print the LQX program generated for solving ``What-If'' models.
.TP
\fB\-q\fP, \fB\-\-queue\-length\fR=\fIarg\fR
Generate gnuplot(1) output to plot the queue length (number of
customers) for either all stations for the class named \fIarg\fP, or
for all classes for the station named \fIarg\fP.  The independent
``What-If'' variable is used as the X-axis.  
.TP
\fB\-Q\fP, \fB\-\-export\-qnap\fR[=\fIstrict\fR]
Convert the input model into QNAP.  If the optional \fIstrict\fR is
present, the conversion \fIwill\fP translate class, station and
variables names into a form suitable for QNAP2 (i.e., no more than
eight characters and no spaces for station and class names).  Spaces
or other non-alphanumeric characters are always converted to
underscores.  The options \fBqrtuw\fP will output QNAP2 print statements to
print out the named results as comma seperated values.
.TP
\fB\-s\fR, \fB\-\-bard\-schweitzer\fR
Use Bard\-Schweitzer approximate MVA to solve the model(s).
.TP
\fB\-r\fP, \fB\-\-response\-time\fR[=\fIarg\fR]
Generate gnuplot(1) output for plotting the response time and bounds for the
system from the solution of a ``What-If'' model.  The response time
does not include the residence times for the reference station.
The independent ``What-If'' variable is used for the X-axis.  If
\fIarg\fP is supplied, only output resutls for the class named \fIarg\fP.
.TP
\fB\-t\fP, \fB\-\-throughput\fR[=\fIarg\fR]
Generate gnuplot(1) output for plotting the throughput.  If the
optional \fIarg\fP is present, the throughput is plotted for all
stations for the class named \fIarg\fP, or for all classes for the
station named \fIarg\fP.  If the optional \fIarg\fP is not present,
generate gnuplot to plot the throughput and throughput bounds for the
``Reference Station'' from the
solution of a ``What-If'' model. The independent ``What-If''
variable is used for the X-axis.  
.IP
If a What-If analysis of population
mix is specified in the input file, then the throughput and bounds for the second
class is plotted versus the throughput and bounds for the first class.
.TP
\fB\-u\fP, \fB\-\-utilization\fR
Generate gnuplot(1) output for plotting the utilization for the for either all stations
for the class named \fIarg\fP, or for all classes for the station
named \fIarg\fP.  The independent ``What-If'' variable is used for the X-axis.
.TP
\fB\-w\fP, \fB\-\-waiting\-time\fR
Generate gnuplot(1) output for plotting the waiting time for either all stations
for the class named \fIarg\fP, or for all classes for the station
named \fIarg\fP.  The independent ``What-If'' variable is used for the X-axis.
.TP
\fB\-X\fP, \fB\-\-debug\-xml\fR
Debug XML input.
.TP
\fB\-X\fP, \fB\-\-debug\-spex\fR
Output the SPEX program generated.
.SH "SEE ALSO"
\fBlqns\fP(1), \fBgnuplot\fP(1).



NEWS
* Bugs fixes in 5.30
208 	lqiolib 	Add --gnuplot output. 
448 	lqiolib 	Faults in open models.
456 	lqiolib 	Add drop probability for open arrivals.
458 	lqiolib 	Schema change for latest version of JMVA.
459 	lqiolib 	LQX output is not saved to a file with -ofilename 
460 	lqiolib 	Multiclass response time plotting is borked. 
384 	lqn2xxx 	Join is permitted in loop
454 	lqn2xxx 	libc++abi: terminating due to uncaught exception of type std::out_of_range: map::at: key not found
432 	lqns 		Bizarre error setting convergence value.
433 	lqns 		51-replication
452 	lqns		56-replication fails. 
449 	petrisrv 	Petrisrvn regression failures
450 	srvndiff 	srvndiff: 34-priority.lqxo:11: error: Invalid argument "-0.538461" to attribute for element <result-task>

* Bugs fixes in 5.29
441     lqiolib         Null model comment.
443     lqiolib         Export/Import the schema "description" field.
445     lqiolib         if ( va_list ) {}
384     lqn2xxx         Join is permitted in loop 
436     lqn2xxx         Model::format() -- bounding box is position sensitive. 
438     lqn2xxx         Queueing model output is not generated.
434     qnsolver        Bounds solution aborts if no whatif found in the input file.
446     qnsolver        Segmentation fault: in `LQX::ObjectPropertyReadNode::invoke() at SyntaxTree.cpp:2096:41

* Bugs fixes in 5.28
413 	Document 	Missing plainurl. 	
205 	lqiolib 	SPEX Convergence variables in XML (Json) files. 
297 	lqiolib 	Gnuplot and result enhancements
359 	lqiolib 	Expressions are not output by Spex::print_header().
411 	lqiolib 	JMVA issues.
416 	lqiolib 	Fixes for plotting population mix WhatIf output. 
422 	lqiolib 	Refactor handling of control variables. 
366 	lqn2xxx 	EXC_BAD_ACCESS Layer::transmorgrify() at layer.cc:571:17 [opt] 
428 	lqn2xxx 	lqn2fig: EXC_BAD_ACCESS libsystem_platform.dylib`_platform_strlen + 4
432 	lqns	 	Bizarre error setting convergence value. 	
433 	lqns	 	51-replication
425 	lqns	 	Refactor initPopulation. 
430 	lqns 		activity-templ.lqn:75: error: Task "Server", and-join "parB & parA" does not match and-fork "parA & parB" at line 74.
418 	lqsim 		lqsim crashes running -t... 	
429 	lqsim 		result.cc:308:12: warning: 'vsprintf' is deprecated: 
356 	miscella 	rpm build issues.
415 	petrisrv 	Signal SIGSEGV in LQIO::DOM::Phase::getServiceTime () at dom_phase.cpp:88 
421 	petrisrv 	SPEX Convergence with PetriSRVN 
423 	petrisrv 	Petrisrvn quorum borked. 
424 	petrisrv 	09-sanity.xlqn 
343 	qnsolver 	Improve compatibility with Java Modelling Tools 
412 	qnsolver 	Save JMVA WhatIf results.
417 	qnsolver 	Bounds solution issues. 
419 	qnsolver 	model.cc:(.text+0x18ff): undefined reference to `typeinfo for QNIO::JMVA_Document'

* Bugs fixed in 5.27
192 	libmva 	 	MVA::marginalProbabilities2 mis-dimensioned. 
401 	libmva 		libc++abi: terminating with uncaught exception of type std::invalid_argument: default-output
403 	libmva 	 	43-activities.lqnx: Floating point exception: mva.cc 541: Inexact result, Invalid operation
406 	libmva 	 	Refactor fpgoop.
383 	lqiolib 	<pre> precedences can accept more than one activity in XML (JSON?) input. 
393 	lqiolib 	Save (and print) marginal queue probabilities.
397 	lqiolib 	Add qnap2 parser for BCMP models.
398 	lqiolib 	Refactor LQX math handing from ExternalVariables to LQX Expressions.
399 	lqiolib 	WhatIf construction is bogus.
411 	lqiolib 	JMVA issues. 
386 	lqn2csv 	getopt_long headaches. 
382 	lqn2xxx 	lqn2ps -Oout doesn't print processors. 
390 	lqn2xxx 	lqn2xxx: command not found. 
400 	lqn2xxx 	Converting from json to lqn fails on Windows.
373 	lqns 	 	Processor waiting times at Rolia multiserver make no sense. 
380 	lqns 	 	lqns hangs with Conway multiserver. 
405 	lqns 	 	Optimizer bug: option.h
407 	lqns 	 	--srvn-layering crashes in V6 interlocking code
391 	lqsim 		lqsim/regression/Makefile.am, 5*-replication. 
356 	miscella 	rpm build issues.
402 	miscella 	Mac M1 issues.
385 	petrisrv 	V5 petrisrvn: faulty models causing aborts.
392 	petrisrv 	Calls are not being constructed
395 	petrisrv 	V5 petrisrvn: faulty models causing aborts or failing.
409 	petrisrv 	Petrisrvn fails when SPEX/LQX is used to alter the arrival rate to an entry. 
387 	srvndiff 	File glob not working. 
389 	srvndiff 	C:\msys64\mingw32\bin\srvndiff.exe: model-001.lqxo:12: error: unclosed token 

* Bugs fixed in 5.26
365 	lqns	 	The warning for zero service time should be suppressed for reference tasks with think time. 
369 	petrisrv 	Segmentation violation with --overtaking 
370 	lqiolib 	Activity parameters probabilities and rates are constants. 
379 	lqns	 	V5 Regression. 	
372 	lqiolib 	Result variables can be reused. 
374 	lqns 		Asynchronous Cycles #1 	
375 	lqns 		Asynchronous Cycles #1 	
376 	lqn2xxx 	Overloaded infinite servers with open arrivals... 
378 	lqn2xxx 	Activity throughput not initialized. 

* Bugs fixed in 5.25
338 	lqns 		Queuing time not equal to zero when NCust = NServers for Zhou multiserver. 
352 	lqn2xxx 	--no-xxxx doesn't work. 
353 	lqns 		Assertion failed: (ix && ix <= sz), function operator[], file vector.h, line 174. 
357 	para_lib 	parasol/include/sem*caphore.h conflicts with /usr/include/semaphore.h 
360 	lqngen	 	lqngen -L<n> doesn't work. 
361 	lqn2xxx 	EXC_BAD_ACCESS (code=EXC_I386_GPFLT) LQIO::DOM::Document::~Document(this=0x0000000100a04080) at dom_document.cpp:123:3 
362 	lqn2csv 	terminate called after throwing an instance of 'std::runtime_error' 
363 	lqn2csv 	vector::_M_range_check: __n (which is 18446744073709551615) >= this->size() (which is 3) 

* Bugs fixed in 5.24
321 	lqiolib 	Runtime Exception Occured: Object of type phase does not have property service_time 
328 	spex 		Using $1, $2... as result variables fails. 
339 	lqiolib 	--reload-lqx segvs or generally is borked. 
340 	lqiolib 	loadResults not implemented in Json_Document. 
341 	lqn2xxx 	lqn2svg -Q1 -o 01-sanity-queue.svg -Zprune 01-sanity.in aborted. 

qnsolver/qnsolver.cc
/*
 * $Id: qnsolver.cc 17169 2024-04-05 23:31:35Z greg $
 */

#include "config.h"
#include <algorithm>
#include <iostream>
#include <fstream>
#include <iomanip>
#include <cstdlib>
#include <cstring>
#include <getopt.h>
#include <libgen.h>
#if HAVE_EXPAT_H
#include <lqio/jmva_document.h>
#endif
#include <lqio/dom_document.h>
#include <lqio/filename.h>
#include <lqio/gnuplot.h>
#include <lqio/qnap2_document.h>
#include "pragma.h"
#include "model.h"

static LQIO::DOM::Pragma pragmas;


static void makeopts( const struct option * longopts, std::string& opts );
static void usage() ;

extern "C" int qnap2debug;

#if HAVE_GETOPT_LONG
const struct option longopts[] =
    /* name */ /* has arg */ /*flag */ /* val */
{
    { "bounds",                                 no_argument,            0, 'b' },
    { "no-bounds",     				no_argument,		0, 0x100+'b' },
    { LQIO::DOM::Pragma::_exact_,               no_argument,            0, 'e' },
    { LQIO::DOM::Pragma::_fast_,                no_argument,            0, 'f' },
    { LQIO::DOM::Pragma::_linearizer_,          no_argument,            0, 'l' },
    { LQIO::DOM::Pragma::_schweitzer_,          no_argument,            0, 's' },
    { LQIO::DOM::Pragma::_force_multiserver_,   no_argument,            0, 'M' },
    { "queue-length",                           optional_argument,      0, 'q' },
    { "response-time",                          optional_argument,      0, 'r' },
    { "throughput",                             optional_argument,      0, 't' },
    { "utilization",                            optional_argument,      0, 'u' },
    { "waiting-time",                           optional_argument,      0, 'w' },
    { "output",                                 required_argument,      0, 'o' },
    { "file-format",                            required_argument,      0, 'O' },
    { "colours",                                required_argument,      0, 'C' },
    { "help",                                   no_argument,            0, 'h' },
    { "multiserver",                            required_argument,      0, 'm' },
    { "no-execute",                             no_argument,            0, 'n' },
    { "verbose",                                no_argument,            0, 'v' },
    { "export-jaba",                            no_argument,            0, 'B' },
    { "export-jmva",                            no_argument,            0, 'J' },
    { "export-qnap",                            optional_argument,      0, 'Q' },
    { "print-lqx",                              no_argument,            0, 'P' },
    { "debug-mva",                              no_argument,            0, 'd' },
    { "debug-qnap2",                            no_argument,            0, 'D' },
    { "debug-lqx",                              no_argument,            0, 'L' },
    { "debug-xml",                              no_argument,            0, 'X' },
    { nullptr, 0, 0, 0 }
};

static std::string opts;
#else
static std::string opts = "bdefhlo:rstvxDGJLQSX";
#endif

const static std::map<const std::string,const std::string> opthelp  = {
    { "bounds",                                 "Use the bounds solver." },
    { "no-bounds",                              "Don't output bounds for response-time or throughout plots." },
    { "colours",				"Use the colours for plotting.  ARG is a list of colours. (not implemented)" },
    { "debug-lqx",                              "Debug the LQX program." },
    { "debug-mva",                              "Enable debug code in the MVA solver." },
    { "debug-qnap2",                            "Debug the QNAP2 input parser." },
    { "debug-xml",                              "Debug XML input." },
    { "export-jmva",				"Export a JMVA model after solution (results embedded)." },
    { "export-jaba",				"Export a JABA model.  Results and reference stations are stripped out." },
    { "export-qnap",                            "Export a QNAP2 model.  Do not solve." },
    { "file-format",				"When plotting, add instructions to output as ARG to the output file." },
    { "help",                                   "Show this." },
    { "multiserver",                            "Use ARG for multiservers.  ARG={conway,reiser,rolia,zhou}." },
    { "no-execute",				"Load the model and run LQX, but do not solve the model." },
    { "output",                                 "Send output to ARG." },
    { "queue-length",                           "Output gnuplot to plot station queue-length.  ARG specifies a class or station." },
    { "response-time",                          "Output gnuplot to plot system response-time (and bounds).  ARG specifies a class." },
    { "throughput",                             "Output gnuplot to plot system throughput (and bounds), or for a class or station with ARG." },
    { "utilization",                            "Output gnuplot to plot utilization.  ARG specifies a class or station." },
    { "waiting-time",                           "Output gnuplot to plot station waiting-times.  ARG specifies a class or station." },
    { "print-lqx",                              "Print the LQX program used to solve the model." },
    { "verbose",                                "" },
    { LQIO::DOM::Pragma::_exact_,               "Use Exact MVA." },
    { LQIO::DOM::Pragma::_fast_,                "Use the Fast Linearizer solver." },
    { LQIO::DOM::Pragma::_force_multiserver_,   "Use the multiserver solution for load independent stations (copies=1)." },
    { LQIO::DOM::Pragma::_linearizer_,          "Use Linearizer." },
    { LQIO::DOM::Pragma::_schweitzer_,          "Use Bard-Schweitzer approximate MVA." },
};

const static std::map<const std::string,const LQIO::GnuPlot::Format> gnuplot_output_format = {
    { "emf",	LQIO::GnuPlot::Format::EMF },
    { "eps",	LQIO::GnuPlot::Format::EPS },
    { "fig",	LQIO::GnuPlot::Format::FIG },
    { "gif",	LQIO::GnuPlot::Format::GIF },
    { "pdf",	LQIO::GnuPlot::Format::PDF },
    { "png",	LQIO::GnuPlot::Format::PNG },
    { "svg",	LQIO::GnuPlot::Format::SVG },
    { "latex",	LQIO::GnuPlot::Format::LATEX }
};

/* Flags */

static bool print_qnap2 = false;		/* Export to qnap2.  		*/
static bool print_jmva = false;			/* Export to JMVA		*/
static bool print_gnuplot = false;		/* Output WhatIf as gnuplot	*/
static BCMP::Model::Result::Type plot_type = BCMP::Model::Result::Type::THROUGHPUT;

/* Globals */

std::string program_name;

#if HAVE_LIBEXPAT
QNIO::JMVA_Document* __input = nullptr;
#endif

/* Local procedures */

static bool exec( const std::string& input_file_name, const std::string& output_file_name, const std::string&, bool, LQIO::GnuPlot::Format );
static bool exec( QNIO::Document& input, const std::string& output_file_name, const std::string&, bool, LQIO::GnuPlot::Format );

int main (int argc, char *argv[])
{
    std::string output_file_name;
    std::string plot_arg;
    LQIO::GnuPlot::Format gnuplot_format = LQIO::GnuPlot::Format::TERMINAL;
    bool no_bounds = false;
    
    program_name = basename( argv[0] );

    /* Process all command line arguments.  If none specified, then     */
#if HAVE_GETOPT_LONG
    makeopts( longopts, opts );
#endif

    LQIO::io_vars.init( VERSION, basename( argv[0] ), LQIO::severity_action );

    pragmas.insert( getenv( "QNSOLVER_PRAGMAS" ) );

    for ( ;; ) {
#if HAVE_GETOPT_LONG
	const int c = getopt_long( argc, argv, opts.c_str(), longopts, NULL );
#else
	const int c = getopt( argc, argv, opts.c_str() );
#endif
	if ( c == EOF ) break;

	switch( c ) {
	case 'b':
	    pragmas.insert(LQIO::DOM::Pragma::_mva_,LQIO::DOM::Pragma::_bounds_);
	    break;

	case 0x100+'b':
	    no_bounds = true;
	    break;
	    
	case 'B':
	    pragmas.insert(LQIO::DOM::Pragma::_mva_,LQIO::DOM::Pragma::_bounds_);
	    break;
	    
	case 'C':
	    std::cerr << "Colours unsupported..." << std::endl;
	    break;

	case 'd':
	    Model::debug_flag = true;
#if DEBUG_MVA
	    MVA::debug_D = true;
	    MVA::debug_L = true;
	    MVA::debug_P = true;
	    MVA::debug_U = true;
	    MVA::debug_W = true;
#endif
	    break;

	case 'D':
	    qnap2debug = 1;
	    break;

	case 'e':
	    pragmas.insert(LQIO::DOM::Pragma::_mva_,LQIO::DOM::Pragma::_exact_);
	    break;

	case 'f':
	    pragmas.insert(LQIO::DOM::Pragma::_mva_,LQIO::DOM::Pragma::_fast_);
	    break;

	case 'h':
	    usage();
	    return 0;

	case 'l':
	    pragmas.insert(LQIO::DOM::Pragma::_mva_,LQIO::DOM::Pragma::_linearizer_);
	    break;

	case 'L':
	    LQIO::DOM::Document::lqx_parser_trace(stderr);
	    break;

	case 'm':
	    pragmas.insert(LQIO::DOM::Pragma::_multiserver_,optarg);
	    break;

	case 'M':
	    pragmas.insert(LQIO::DOM::Pragma::_force_multiserver_,LQIO::DOM::Pragma::_true_);
	    break;

	case 'n':
	    Model::no_execute = true;
	    break;
	    
	case 'o':
            output_file_name = optarg;
	    break;

	case 'O':
	    if ( optarg != nullptr ) {
		std::map<const std::string,const LQIO::GnuPlot::Format>::const_iterator format = gnuplot_output_format.find( optarg );
		if ( format == gnuplot_output_format.end() ) {
		    std::cerr << program_name << ": Unsupported gnuplot format: " << optarg << "." << std::endl;
		} else {
		    gnuplot_format = format->second;
		}
	    }
	    break;
	    
	case 'q':
	    print_gnuplot = true;			/* Output WhatIf as gnuplot	*/
	    plot_type = BCMP::Model::Result::Type::QUEUE_LENGTH;
	    if ( optarg != nullptr ) plot_arg = optarg;
	    break;

	case 'r':
	    print_gnuplot = true;			/* Output WhatIf as gnuplot	*/
	    plot_type = BCMP::Model::Result::Type::RESPONSE_TIME;
	    if ( optarg != nullptr ) plot_arg = optarg;
	    break;

	case 's':
	    pragmas.insert(LQIO::DOM::Pragma::_mva_,LQIO::DOM::Pragma::_schweitzer_);
	    break;

	case 't':
	    print_gnuplot = true;			/* Output WhatIf as gnuplot	*/
	    plot_type = BCMP::Model::Result::Type::THROUGHPUT;
	    if ( optarg != nullptr ) plot_arg = optarg;
	    break;

	case 'u':
	    print_gnuplot = true;			/* Output WhatIf as gnuplot	*/
	    plot_type = BCMP::Model::Result::Type::UTILIZATION;
	    if ( optarg != nullptr ) plot_arg = optarg;
	    break;

	case 'v':
	    Model::verbose_flag = true;
	    break;

	case 'w':
	    print_gnuplot = true;			/* Output WhatIf as gnuplot	*/
	    plot_type = BCMP::Model::Result::Type::RESIDENCE_TIME;
	    if ( optarg != nullptr ) plot_arg = optarg;
	    break;

	case 'x':
	    /* Not implemented */
	    break;

	case 'J':
	    print_jmva = true;
	    break;
	    
	case 'Q':
	    print_qnap2 = true;
	    if ( optarg != nullptr ) {
		if ( strcmp(optarg,"strict") == 0 ) {
		    QNIO::QNAP2_Document::__strict = true;
		} else {
		    std::cerr << program_name << ": --print-qnap=" << optarg << ", unknow argument." << std::endl;
		}
	    } 
	    break;

	case 'P':
	    Model::print_program = true;
	    break;

	case 'X':
            LQIO::DOM::Document::__debugXML = true;
	    break;

	default:
	    usage();
	    return 1;
	}
    }

    /* input is assumed to come in from stdin.                          */

    int rc = 0;
#if HAVE_EXPAT_H
    if ( optind == argc ) {
	rc = exec( "-", output_file_name, plot_arg, no_bounds, gnuplot_format ) ? 0 : 1;
    } else {
        for ( ; optind < argc; ++optind ) {
	    rc |= (exec( argv[optind], output_file_name, plot_arg, no_bounds, gnuplot_format ) ? 0 : 1);
	}
    }
#else
    std::cerr << LQIO::io_vars.lq_toolname << ": No expat library available." << std::endl;
#endif
    return rc;
}


/*
 * Run the solver.
 */

static bool exec( const std::string& input_file_name, const std::string& output_file_name, const std::string& plot_arg, bool no_bounds, LQIO::GnuPlot::Format plot_format )
{
    bool rc = false;
    if ( Model::verbose_flag ) std::cerr << input_file_name << ": load... ";
    if ( LQIO::DOM::Document::getInputFormatFromFilename( input_file_name, LQIO::DOM::Document::InputFormat::JMVA ) == LQIO::DOM::Document::InputFormat::QNAP2 ) {
	QNIO::QNAP2_Document input( input_file_name );
	rc = exec( input, output_file_name, plot_arg, no_bounds, plot_format );
#if HAVE_LIBEXPAT
    } else {
	QNIO::JMVA_Document input( input_file_name );
	rc = exec( input, output_file_name, plot_arg, no_bounds, plot_format );
#endif
    }
    if ( Model::verbose_flag ) std::cerr << "done" << std::endl;
    return rc;
}


static bool exec( QNIO::Document& input, const std::string& output_file_name, const std::string& plot_arg, bool no_bounds, LQIO::GnuPlot::Format plot_format )
{
    if ( !input.load() || LQIO::io_vars.error_count > 0 ) return false;
    input.mergePragmas( pragmas.getList() );
    Pragma::set( input.getPragmaList() );		/* load pragmas here */

    const bool qnap_model = input.getInputFormat() == QNIO::Document::InputFormat::QNAP;
    
    if ( print_gnuplot && qnap_model ) {
	std::cerr << LQIO::io_vars.lq_toolname << ": plotting not supported with QNAP input." << std::endl;
    }
    std::ofstream output;
    const bool bounds = Pragma::mva() == Model::Solver::BOUNDS || input.boundsOnly();
    if ( !output_file_name.empty() ) {
	output.open( output_file_name, std::ios::out );
	if ( !output ) {
	    std::cerr << LQIO::io_vars.lq_toolname << ": Cannot open output file \"" << output_file_name << "\" -- " << strerror( errno ) << std::endl;
	    return false;
	}
    } else if ( print_jmva ) {
	const std::string extension = bounds ? "jmva" : "jaba";
	LQIO::Filename filename( input.getInputFileName(), extension );
	LQIO::Filename::backup( filename() );
	output.open( filename(), std::ios::out );
	if ( !output ) {
	    std::cerr << LQIO::io_vars.lq_toolname << ": Cannot open output file \"" << input.getInputFileName() << "\" -- " << strerror( errno ) << std::endl;
	    return false;
	}
    }

    if ( print_qnap2 ) {

	QNIO::QNAP2_Document qnap_model( input );
	try {
	    if ( print_gnuplot ) qnap_model.plot( plot_type, plot_arg, no_bounds, plot_format );
	}
	catch ( const std::invalid_argument& e ) {
	    std::cerr << LQIO::io_vars.lq_toolname << ": Invalid class or station name for --plot: " << e.what() << std::endl;
	}
	if ( output_file_name.empty() ) {
	    LQIO::Filename filename( input.getInputFileName(), "qnap" );
	    LQIO::Filename::backup( filename() );
	    output.open( filename(), std::ios::out );
	    if ( !output ) {
		std::cerr << LQIO::io_vars.lq_toolname << ": Cannot open output file \"" << input.getInputFileName() << "\" -- " << strerror( errno ) << std::endl;
		return false;
	    }
	} 
	qnap_model.exportModel( output );

    } else {
	try {
	    if ( print_gnuplot ) input.plot( plot_type, plot_arg, no_bounds, plot_format );
	}
	catch ( const std::invalid_argument& e ) {
	    std::cerr << LQIO::io_vars.lq_toolname << ": Invalid class or station name for --plot: " << e.what() << std::endl;
	}
	if ( print_jmva ) {
	    /* Since we might get QNAP in, rexport directly -- WhatIf?? */
	    Model model( input, Pragma::mva(), std::string() );
	    model.solve();
//	    QNIO::JMVA_Document new_model( output_file_name, input.model() );
//	    new_model.comprehensions() = input.comprehensions();
//	    new_model.exportModel( output );	/* Will save all results (if !bounds) */
	    input.exportModel( output );
	} else {
	    Model model( input, Pragma::mva(), output_file_name );
	    model.solve();
	}
    }
    if ( output ) {
	output.close();
    }
    return true;
}

static void
makeopts( const struct option * longopts, std::string& opts )
{
    for ( int i = 0; longopts[i].name != 0; ++i ) {
	if ( isgraph( longopts[i].val ) ) {
	    opts += longopts[i].val;
	    if ( longopts[i].has_arg == required_argument ) {
		opts += ':';
	    }
	}
    }
}

static void
usage()
{
    std::cerr << "Usage: " << program_name;

#if HAVE_GETOPT_LONG
    std::cerr << " [option]" << std::endl << std::endl;
    std::cerr << "Options" << std::endl;
    for ( const struct option *o = longopts; (o->name || o->val); ++o ) {
	std::string s;
	if ( o->name ) {
	    s = "--";
	    s += o->name;
	    if ( o->has_arg != no_argument ) {
		s += "=ARG";
	    }
	} else {
	    s = " ";
	}
	if ( isascii(o->val) && isgraph(o->val) ) {
	    std::cerr << " -" << static_cast<char>(o->val) << ", ";
	} else {
	    std::cerr << "     ";
	}
	std::cerr.setf( std::ios::left, std::ios::adjustfield );
	std::cerr << std::setw(28) << s << opthelp.at(o->name) << std::endl;
    }
#else
    const char * s;
    std::cerr << " [-";
    for ( s = opts.c_str(); *s; ++s ) {
	if ( *(s+1) == ':' ) {
	    ++s;
	} else {
	    std::cerr.put( *s );
	}
    }
    std::cerr << ']';

    for ( s = opts.c_str(); *s; ++s ) {
	if ( *(s+1) == ':' ) {
	    std::cerr << " [-" << *s;
	    switch ( *s ) {
	    }
	    std::cerr << ']';
	    ++s;
	}
    }
#endif
    std::cerr << std::endl;
}



/*
 * What to do based on the severity of the error.
 */

void
LQIO::severity_action (LQIO::error_severity severity)
{
    switch( severity ) {
    case LQIO::error_severity::FATAL:
	(void) abort();
	break;

    case LQIO::error_severity::ERROR:
	LQIO::io_vars.error_count += 1;
	if  ( LQIO::io_vars.error_count >= LQIO::io_vars.max_error ) {
	    throw std::runtime_error( "Too many errors" );
	}
	break;
    default:;
    }
}
